miscellaneous_scripts/extract_apt_patch_data.py

```diff
@@ -0,0 +1,172 @@
+    result = otpt.stdout.strip().split("\n")
+    package_dict = {}
+    for output_line in result:
+        if "/" not in output_line:
+            continue
+        package_name = output_line.split("/")[0]
+        if package_name not in package_list:
+            continue
+        if package_name in package_dict:
+            raise ValueError(f"Package {package_name} already exists in {package_dict}")
+        package_dict[package_name] = {}
+        installed_version = output_line.split()[1]
+        package_dict[package_name]["installed_version"] = installed_version
+    return package_dict
+
+
+def process_packages(
+    installed_packages,
+    impacted_packages,
+    upgradable_packages,
+    patch_package_list,
+    upgradable_packages_data_for_impacted_packages,
+):
+    """
+    This method handles the processing of the apt packages.  The ECR Enhanced Scan gives the info about the source
+    package that is impacted, but does not give the exact package/binary that is impacted. Hence, we need to find out
+    the source of each package and see if the source or the package is impacted. If either is impacted and the package
+    is upgradable, it is stored in the patch_package_list. The patch_package_list consists of all the packages that need
+    to be upgraded and their current version. It also updates a dictionary called upgradable_packages_data_for_impacted_packages.
+    This dict maintains the source package names as the keys and all their packages/binaries as values.
+
+    :param installed_packages: list, List of all the installed apt packages
+    :param impacted_packages: list, List of all the impacted apt packages (or source apt packages)
+    :param upgradable_packages: list, List of all the upgradable apt packages
+    :param patch_package_list: list, List of all the apt packages that need to be upgraded for patching. One of the core functionalities of this
+                                     method is to add data to this list.
+    :param upgradable_packages_data_for_impacted_packages: dict[List], Dict of all the impacted source apt packages as keys and a list of their upgradable binaries as value.
+                                                                       One of the core functionalities of this method is to add data to this dict.
+    :return patch_package_list: list, the same input parameter that is modified by this function
+    :return upgradable_packages_data_for_impacted_packages: dict[list], the same input parameter that is modified by this function
+    """
+    for package in installed_packages:
+        source_package = ""
+        source_extraction_cmd_output = subprocess.run(
+            f"dpkg -s {package} | grep ^Source", shell=True, capture_output=True, text=True
+        )
+        if source_extraction_cmd_output.returncode in RETURN_CODE_OK:
+            source_package = source_extraction_cmd_output.stdout.strip().split()[1]
+
+        if (
```

Comment:
If this not being used in any other part of the script, add a comment. If it is being used, make a helper function

Reply:
Made separate method - done